fendermint/testing/contract-test/tests/upgrades.rs

```diff
@@ -0,0 +1,300 @@
+use fendermint_vm_interpreter::GenesisInterpreter;
+use fendermint_vm_interpreter::{
+    fvm::{bundle::contracts_path, upgrade_scheduler::Upgrade, FvmMessageInterpreter},
+    ExecInterpreter,
+};
+
+#[derive(Clone)]
+struct Tester<I> {
+    interpreter: Arc<I>,
+    state_store: Arc<MemoryBlockstore>,
+    multi_engine: Arc<MultiEngine>,
+    exec_state: Arc<tokio::sync::Mutex<Option<FvmExecState<MemoryBlockstore>>>>,
+    state_params: FvmStateParams,
+}
+
+impl<I> Tester<I>
+where
+    I: GenesisInterpreter<
+        State = FvmGenesisState<MemoryBlockstore>,
+        Genesis = Genesis,
+        Output = FvmGenesisOutput,
+    >,
+    I: ExecInterpreter<
+        State = FvmExecState<MemoryBlockstore>,
+        Message = FvmMessage,
+        BeginOutput = FvmApplyRet,
+        DeliverOutput = FvmApplyRet,
+        EndOutput = PowerUpdates,
+    >,
+{
+    fn state_store_clone(&self) -> MemoryBlockstore {
+        self.state_store.as_ref().clone()
+    }
+
+    pub fn new(interpreter: I, state_store: MemoryBlockstore) -> Self {
+        Self {
+            interpreter: Arc::new(interpreter),
+            state_store: Arc::new(state_store),
+            multi_engine: Arc::new(MultiEngine::new(1)),
+            exec_state: Arc::new(tokio::sync::Mutex::new(None)),
+            state_params: FvmStateParams {
+                timestamp: Timestamp(0),
+                state_root: Cid::default(),
+                network_version: NetworkVersion::V21,
+                base_fee: TokenAmount::zero(),
+                circ_supply: TokenAmount::zero(),
+                chain_id: 0,
+                power_scale: 0,
+            },
+        }
+    }
+
+    async fn init_genesis(&mut self) -> anyhow::Result<()> {
+        let bundle_path = bundle_path();
+        let bundle = std::fs::read(&bundle_path)
+            .with_context(|| format!("failed to read bundle: {}", bundle_path.to_string_lossy()))?;
+
+        let custom_actors_bundle_path = custom_actors_bundle_path();
+        let custom_actors_bundle =
+            std::fs::read(&custom_actors_bundle_path).with_context(|| {
+                format!(
+                    "failed to read custom actors_bundle: {}",
+                    custom_actors_bundle_path.to_string_lossy()
+                )
+            })?;
+
+        let state = FvmGenesisState::new(
+            self.state_store_clone(),
+            self.multi_engine.clone(),
+            &bundle,
+            &custom_actors_bundle,
+        )
+        .await
+        .context("failed to create genesis state")?;
+
+        let genesis = Genesis {
+            chain_name: "test".to_string(),
+            timestamp: Timestamp(0),
+            network_version: NetworkVersion::V21,
+            base_fee: TokenAmount::zero(),
+            power_scale: 0,
+            validators: Vec::new(),
+            accounts: Vec::new(),
+            eam_permission_mode: PermissionMode::Unrestricted,
+            ipc: None,
```

Comment:
Note that if this is `None` then IPC actors won't be deployed. That is fine, and another reason to pass in genesis from the test, so this `Tester` can be as general as possible, with no hidden behaviour.

Reply:
exactly, right now I am initializing part of the genesis from outside the tester (accounts), will try to make it as generic as possible